### ISSUE TITLE
net/http: Expose the normal transport RoundTripper for WASM/js

### ISSUE THREAD

**Issue by antong** [golang/go#27495]
On WASM, the standard `http.Transfer` RoundTripper is not exposed for external use. Buildflags for wasm replace the standard implementation of RoundTripper with a special version that uses the browser Fetch API. The special JS RoundTripper does not seem to use or honor any of the documented settings provided by `http.Transport`, such as `Dial` or `DialContext`. But the way the special JS RoundTripper is implemented as methods on `http.Transport`, it prevents using the standard RoundTripper.

There are use cases for using the normal RoundTripper, for instance for testing and simulation in the browser. The normal RoundTripper is compiled and included in the binary anyway, and works perfectly well, but is masked by the browser version.

### What version of Go are you using (`go version`)? 

go1.11

### What operating system and processor architecture are you using (`go env`)?

GOOS=js
GOARCH=wasm

### What did you do?

I tried to configure a `http.Client` to use a transport with a custom dial function:
https://play.golang.org/p/dnyTivQ643J

This works on all other platforms including the playground, but on WASM, the same, special Roundtripper is always used and there is no way to use the standard Roundtripper.

### What did you expect to see?

```
Get https://api.ipify.org/: expected error, all ok
```

### What did you see instead?

On WASM:
```
192.0.2.77
Done.
```

On other platforms, I get the expected result

**Comment by andybons** [comment]
@neelance @dmitshur

**Comment by antong** [comment]
I quickly tested two hacks to do this, and both seemed to work. Both pass the tests on normal platforms, but I don't know how to run the unit tests on WASM.

Hack 1: Don't implement the special RoundTripper on `http.Transport`, but make an own jsTransport. Set an instance of this as `http.DefaultTransport` for GOARCH=wasm. Example: e8050da

Hack 2: Mod the special RoundTripper to instead use the normal RoundTripper if the caller has set options on the Transport. Example: 8137bf5

**Comment by agnivade** [comment]
@johanbrandhorst

**Comment by rubensayshi** [comment]
running into this problem when trying to do HTTPS requests using self signed certs over a websocket connection.

https://github.com/golang/go/commit/e8050da2dd93f4ff00a590c14f94c31da3c3159b seems like an easy and clean solution to enable a lot more flexibility when desired?

**Comment by johanbrandhorst** [comment]
I agree Hack 1 as proposed by @antong is a reasonable short term approach. Are you interested in contributing this?

**Comment by antong** [comment]
Sure! Do you mean that I should make a CL based on Hack 1 (e8050da)?

**Comment by johanbrandhorst** [comment]
Yes, that sounds good.

**Comment by hullarb** [comment]
Hi all, 

i just bumped into this issues as i tried to implement some tunneling for http request in wasm.
It's a pity that the above PR was abandoned.
i also have a very minimalistic change( https://github.com/golang/go/commit/8327844bf608a69e45f36c21fb804b456d6d8e8c) that i could use for testing my http tunneling.

As Transport currently does not provide any ways to customize the wasm/js RoundTrip behavior i wonder if it would not be fair to simple fallback to the "normal" roundTrip implementation in case a custom Transport struct is the caller and not the DeafultTransport (
the js RoundTrip implementation has already a fallback to the real transport RoundTrip implementation in case the fetch api is not available).
